File: app/src/main/java/fr/insapp/insapp/activities/SearchActivity.java
package fr.insapp.insapp.activities;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.widget.LinearLayout;
import android.widget.Toast;

import java.util.Calendar;
import java.util.Date;

import fr.insapp.insapp.R;
import fr.insapp.insapp.adapters.ClubRecyclerViewAdapter;
import fr.insapp.insapp.adapters.EventRecyclerViewAdapter;
import fr.insapp.insapp.adapters.PostRecyclerViewAdapter;
import fr.insapp.insapp.adapters.UserRecyclerViewAdapter;
import fr.insapp.insapp.http.ServiceGenerator;
import fr.insapp.insapp.models.Club;
import fr.insapp.insapp.models.Event;
import fr.insapp.insapp.models.Post;
import fr.insapp.insapp.models.SearchTerms;
import fr.insapp.insapp.models.UniversalSearchResults;
import fr.insapp.insapp.models.User;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by thomas on 11/12/2016.
 */

public class SearchActivity extends AppCompatActivity {

    private ClubRecyclerViewAdapter adapterClubs;
    private PostRecyclerViewAdapter adapterPosts;
    private EventRecyclerViewAdapter adapterEvents;
    private UserRecyclerViewAdapter adapterUsers;

    private String query;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        // query

        Intent intent = getIntent();
        if (Intent.ACTION_SEARCH.equals(intent.getAction())) {
            this.query = intent.getStringExtra(SearchManager.QUERY);
        }

        // toolbar

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_search);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            getSupportActionBar().setHomeButtonEnabled(true);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        // clubs recycler view

        RecyclerView recyclerViewClubs = (RecyclerView) findViewById(R.id.recyclerview_search_clubs);
        recyclerViewClubs.setHasFixedSize(true);
        recyclerViewClubs.setNestedScrollingEnabled(false);

        LinearLayoutManager layoutManagerClubs = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        recyclerViewClubs.setLayoutManager(layoutManagerClubs);

        this.adapterClubs = new ClubRecyclerViewAdapter(this, false);
        adapterClubs.setOnItemClickListener(new ClubRecyclerViewAdapter.OnClubItemClickListener() {
            @Override
            public void onClubItemClick(Club club) {
                startActivity(new Intent(getBaseContext(), ClubActivity.class).putExtra("club", club));
            }
        });

        recyclerViewClubs.setAdapter(adapterClubs);

        // posts recycler view

        RecyclerView recyclerViewPosts = (RecyclerView) findViewById(R.id.recyclerview_search_posts);
        recyclerViewPosts.setHasFixedSize(true);
        recyclerViewPosts.setNestedScrollingEnabled(false);

        LinearLayoutManager layoutManagerPosts = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        recyclerViewPosts.setLayoutManager(layoutManagerPosts);

        this.adapterPosts = new PostRecyclerViewAdapter(this, R.layout.row_post);
        adapterPosts.setOnItemClickListener(new PostRecyclerViewAdapter.OnPostItemClickListener() {
            @Override
            public void onPostItemClick(Post post) {
                startActivity(new Intent(getBaseContext(), PostActivity.class).putExtra("post", post));
            }
        });

        recyclerViewPosts.setAdapter(adapterPosts);

        // events recycler view

        RecyclerView recyclerViewEvents = (RecyclerView) findViewById(R.id.recyclerview_search_events);
        recyclerViewEvents.setHasFixedSize(true);
        recyclerViewEvents.setNestedScrollingEnabled(false);

        LinearLayoutManager layoutManagerEvents = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        recyclerViewEvents.setLayoutManager(layoutManagerEvents);

        this.adapterEvents = new EventRecyclerViewAdapter(this, R.layout.row_event_with_avatars);
        adapterEvents.setOnItemClickListener(new EventRecyclerViewAdapter.OnEventItemClickListener() {
            @Override
            public void onEventItemClick(Event event) {
                startActivity(new Intent(getBaseContext(), EventActivity.class).putExtra("event", event));
            }
        });
        recyclerViewEvents.setAdapter(adapterEvents);

        // users recycler view

        RecyclerView recyclerViewUsers = (RecyclerView) findViewById(R.id.recyclerview_search_users);
        recyclerViewUsers.setHasFixedSize(true);
        recyclerViewUsers.setNestedScrollingEnabled(false);

        LinearLayoutManager layoutManagerUsers = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        recyclerViewUsers.setLayoutManager(layoutManagerUsers);

        this.adapterUsers = new UserRecyclerViewAdapter(this, false);
        adapterUsers.setOnItemClickListener(new UserRecyclerViewAdapter.OnUserItemClickListener() {
            @Override
            public void onUserItemClick(User user) {
                startActivity(new Intent(getBaseContext(), ProfileActivity.class).putExtra("user", user));
            }
        });

        recyclerViewUsers.setAdapter(adapterUsers);

        // hide layouts

        findViewById(R.id.search_clubs_layout).setVisibility(LinearLayout.GONE);
        findViewById(R.id.search_posts_layout).setVisibility(LinearLayout.GONE);
        findViewById(R.id.search_events_layout).setVisibility(LinearLayout.GONE);
        findViewById(R.id.search_users_layout).setVisibility(LinearLayout.GONE);

        // search

        generateSearchResults();
    }

    private void generateSearchResults() {
        if (this.query != null) {
            adapterClubs.getClubs().clear();
            adapterPosts.getPosts().clear();
            adapterEvents.getEvents().clear();
            adapterUsers.getUsers().clear();

            Call<UniversalSearchResults> call = ServiceGenerator.create().universalSearch(new SearchTerms(query));
            call.enqueue(new Callback<UniversalSearchResults>() {
                @Override
                public void onResponse(@NonNull Call<UniversalSearchResults> call, @NonNull Response<UniversalSearchResults> response) {
                    if (response.isSuccessful()) {
                        final UniversalSearchResults results = response.body();

                        if (results.getClubs() != null) {
                            for (int i = 0; i < results.getClubs().size(); i++) {
                                final Club club = results.getClubs().get(i);

                                if (!club.getProfilPicture().isEmpty() && !club.getCover().isEmpty()) {
                                    adapterClubs.addItem(club);
                                    findViewById(R.id.search_clubs_layout).setVisibility(LinearLayout.VISIBLE);
                                }
                            }
                        }

                        if (results.getPosts() != null) {
                            for (int i = 0; i < results.getPosts().size(); i++) {
                                adapterPosts.addItem(results.getPosts().get(i));
                                findViewById(R.id.search_posts_layout).setVisibility(LinearLayout.VISIBLE);
                            }
                        }

                        if (results.getEvents() != null) {
                            final Date atm = Calendar.getInstance().getTime();

                            for (int i = 0; i < results.getEvents().size(); i++) {
                                final Event event = results.getEvents().get(i);

                                if (event.getDateEnd().getTime() > atm.getTime()) {
                                    adapterEvents.addItem(event);
                                    findViewById(R.id.search_events_layout).setVisibility(LinearLayout.VISIBLE);
                                }
                            }
                        }

                        if (results.getUsers() != null) {
                            for (int i = 0; i < results.getUsers().size(); i++) {
                                adapterUsers.addItem(results.getUsers().get(i));
                                findViewById(R.id.search_users_layout).setVisibility(LinearLayout.VISIBLE);
                            }
                        }
                    }
                    else {
                        Toast.makeText(SearchActivity.this, "SearchActivity", Toast.LENGTH_LONG).show();
                    }
                }

                @Override
                public void onFailure(@NonNull Call<UniversalSearchResults> call, @NonNull Throwable t) {
                    Toast.makeText(SearchActivity.this, "SearchActivity", Toast.LENGTH_LONG).show();
                }
            });
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }
}

File: app/src/main/java/fr/insapp/insapp/activities/SearchActivity.kt
package fr.insapp.insapp.activities

import android.app.SearchManager
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.widget.LinearLayout
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.SearchEvent
import fr.insapp.insapp.R
import fr.insapp.insapp.adapters.AssociationRecyclerViewAdapter
import fr.insapp.insapp.adapters.EventRecyclerViewAdapter
import fr.insapp.insapp.adapters.PostRecyclerViewAdapter
import fr.insapp.insapp.adapters.UserRecyclerViewAdapter
import fr.insapp.insapp.http.ServiceGenerator
import fr.insapp.insapp.models.SearchTerms
import fr.insapp.insapp.models.UniversalSearchResults
import kotlinx.android.synthetic.main.activity_search.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

/**
 * Created by thomas on 11/12/2016.
 */

class SearchActivity : AppCompatActivity() {

    private lateinit var adapterClubs: AssociationRecyclerViewAdapter
    private lateinit var adapterPosts: PostRecyclerViewAdapter
    private lateinit var adapterEvents: EventRecyclerViewAdapter
    private lateinit var adapterUsers: UserRecyclerViewAdapter

    private var query: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        // query

        val intent = intent
        if (Intent.ACTION_SEARCH == intent.action) {
            this.query = intent.getStringExtra(SearchManager.QUERY)

            // Answers

            Answers.getInstance().logSearch(SearchEvent().putQuery(query))
        }

        // toolbar

        setSupportActionBar(toolbar_search)
        supportActionBar?.setHomeButtonEnabled(true)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val requestManager = Glide.with(this)

        // clubs recycler view

        recyclerview_search_clubs.setHasFixedSize(true)
        recyclerview_search_clubs.isNestedScrollingEnabled = false

        val layoutManagerClubs = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        recyclerview_search_clubs.layoutManager = layoutManagerClubs

        this.adapterClubs = AssociationRecyclerViewAdapter(mutableListOf(), requestManager, false)
        recyclerview_search_clubs.adapter = adapterClubs

        // posts recycler view

        recyclerview_search_posts.setHasFixedSize(true)
        recyclerview_search_posts.isNestedScrollingEnabled = false

        val layoutManagerPosts = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerview_search_posts.layoutManager = layoutManagerPosts

        this.adapterPosts = PostRecyclerViewAdapter(mutableListOf(), requestManager, R.layout.row_post)
        recyclerview_search_posts.adapter = adapterPosts

        // events recycler view

        recyclerview_search_events.setHasFixedSize(true)
        recyclerview_search_events.isNestedScrollingEnabled = false

        val layoutManagerEvents = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        recyclerview_search_events.layoutManager = layoutManagerEvents

        this.adapterEvents = EventRecyclerViewAdapter(mutableListOf(), requestManager, false, R.layout.row_event_with_avatars)
        recyclerview_search_events.adapter = adapterEvents

        // users recycler view

        recyclerview_search_users.setHasFixedSize(true)
        recyclerview_search_users.isNestedScrollingEnabled = false

        val layoutManagerUsers = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        recyclerview_search_users.layoutManager = layoutManagerUsers

        this.adapterUsers = UserRecyclerViewAdapter(mutableListOf(), requestManager, false)
        recyclerview_search_users.adapter = adapterUsers

        // hide layouts

        search_clubs_layout?.visibility = LinearLayout.GONE
        search_posts_layout?.visibility  = LinearLayout.GONE
        search_events_layout?.visibility  = LinearLayout.GONE
        search_users_layout?.visibility = LinearLayout.GONE
        search_no_result?.visibility = LinearLayout.GONE

        // search

        generateSearchResults()
    }

    private fun generateSearchResults() {
        if (this.query != null) {
            adapterClubs.associations.clear()
            adapterPosts.posts.clear()
            adapterEvents.events.clear()
            adapterUsers.users.clear()

            val call = ServiceGenerator.create().universalSearch(SearchTerms(query))
            call.enqueue(object : Callback<UniversalSearchResults> {
                override fun onResponse(call: Call<UniversalSearchResults>, response: Response<UniversalSearchResults>) {
                    if (response.isSuccessful) {
                        val results = response.body()

                        if (results!!.clubs != null) {
                            for (i in 0 until results.clubs.size) {
                                val club = results.clubs[i]

                                if (club.profilePicture.isNotEmpty() && club.cover.isNotEmpty()) {
                                    adapterClubs.addItem(club)
                                    search_clubs_layout.visibility = LinearLayout.VISIBLE
                                }
                            }
                        }

                        if (results.posts != null) {
                            for (i in 0 until results.posts.size) {
                                adapterPosts.addItem(results.posts[i])
                                search_posts_layout.visibility = LinearLayout.VISIBLE
                            }
                        }

                        if (results.events != null) {
                            val atm = Calendar.getInstance().time

                            for (i in 0 until results.events.size) {
                                val event = results.events[i]

                                if (event.dateEnd.time > atm.time) {
                                    adapterEvents.addItem(event)
                                    search_events_layout.visibility = LinearLayout.VISIBLE
                                }
                            }
                        }

                        if (results.users != null) {
                            for (i in 0 until results.users.size) {
                                adapterUsers.addItem(results.users[i])
                                search_users_layout.visibility = LinearLayout.VISIBLE
                            }
                        }

                        if(results.clubs == null && results.events == null && results.posts == null && results.users == null){
                            search_no_result?.visibility = LinearLayout.VISIBLE
                        }

                    } else {
                        Toast.makeText(this@SearchActivity, "SearchActivity", Toast.LENGTH_LONG).show()
                    }
                }

                override fun onFailure(call: Call<UniversalSearchResults>, t: Throwable) {
                    Toast.makeText(this@SearchActivity, "SearchActivity", Toast.LENGTH_LONG).show()
                }
            })
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }
}